src/openrct2/scripting/ScEntity.hpp

```diff
@@ -1057,6 +1057,251 @@ namespace OpenRCT2::Scripting
         }
     };
 
+    class ScPatrolArea
+    {
+    public:
+        ScPatrolArea(uint16_t staffId, std::vector<MapRange> regions)
+            : _staffId(staffId)
+            , _regions(regions)
+        {
+            mergeRegions();
+        }
+
+        static void Register(duk_context* ctx)
+        {
+            dukglue_register_property(ctx, &ScPatrolArea::staffId_get, nullptr, "staffId");
+            dukglue_register_property(ctx, &ScPatrolArea::regions_get, nullptr, "regions");
+            dukglue_register_method(ctx, &ScPatrolArea::add, "add");
+            dukglue_register_method(ctx, &ScPatrolArea::remove, "remove");
+            dukglue_register_method(ctx, &ScPatrolArea::contains, "contains");
+        }
+
+    private:
+
+        static inline int32_t patrolRegionHeight(const MapRange& region)
+        {
+            return (region.GetY2() - region.GetY1()) / (COORDS_XY_STEP * 4);
+        }
+
+        static inline int32_t patrolRegionWidth(const MapRange& region)
+        {
+            return (region.GetX2() - region.GetX1()) / (COORDS_XY_STEP * 4);
+        }
+
+        static inline void mergeRegions(MapRange& target, const MapRange& toMerge)
+        {
+            //printf(
+            //    "target %d %d %d %d\n", target.Point1.x / 128, target.Point1.y / 128, target.Point2.x / 128,
+            //    target.Point2.y / 128);
+            //printf(
+            //    "toMerge %d %d %d %d\n", toMerge.Point1.x / 128, toMerge.Point1.y / 128, toMerge.Point2.x / 128,
+            //    toMerge.Point2.y / 128);
+            target.Point1.x = std::min(target.Point1.x, toMerge.Point1.x);
+            target.Point1.y = std::min(target.Point1.y, toMerge.Point1.y);
+            target.Point2.x = std::max(target.Point2.x, toMerge.Point2.x);
+            target.Point2.y = std::max(target.Point2.y, toMerge.Point2.y);
+            //printf(
+            //    "merged %d %d %d %d\n\n", target.Point1.x / 128, target.Point1.y / 128, target.Point2.x / 128,
+            //    target.Point2.y / 128);
+        }
+
+        void mergeRegions()
+        {
+            // from top left, go across horizontal scanlines to create as many one
+            // horizontally maximized regions; then go through vertical scanlines and merge
+            // as many of the horizontal regions as possible into bigger rectangles; should
+            // produce minimum possible number of rectangular regions (though not
+            // necessarily the most "square" ones)
+            constexpr int32_t PATROL_REGION_STEP = COORDS_XY_STEP * COORDS_XY_PATROL_STEP;
+            std::array<std::array<MapRange*, 64>, 64> patrolBitmask = { nullptr };
```

Comment:
quite a large stack allocation; could be replaced with a `std::vector<std::vector<MapRange*>>` and populated in place on line 1138 below